massa-bootstrap/src/server.rs

```diff
@@ -14,363 +42,451 @@ use massa_signature::KeyPair;
+            // - 3.b. If present, fall-back to the stop behaviour
+            // - 3.c. If absent, all's clear to rock-n-roll.
+
+            // 1. Block until _something_ is ready
+            let rdy = selector.ready();
+
+            // 2. If that something is a stop-signal, stop
+            // from `crossbeam::Select::read()` documentation:
+            // "Note that this method might return with success spuriously, so it’s a good idea
+            // to always double check if the operation is really ready."
+            if rdy == 0 && self.stopper_rx.try_recv().is_ok() {
+                break;
+                // - 3. If that something is anything else:
+            } else {
+                massa_trace!("bootstrap.lib.run.select", {});
+
+                // - 3.a. double check the stop-signal is absent
+                let stop = self.stopper_rx.try_recv();
+                if unlikely(stop.is_ok()) {
                     massa_trace!("bootstrap.lib.run.select.manager", {});
-                    break
+                    break;
+                } else if unlikely(stop == Err(crossbeam::channel::TryRecvError::Disconnected)) {
+                    return Err(BootstrapError::GeneralError(
+                        "Unexpected stop-channel disconnection".to_string(),
+                    )
+                    .into());
+                };
+            }
+            // - 3.c. If absent, all's clear to rock-n-roll.
+            let (dplx, remote_addr) = match self.listener_rx.try_recv() {
+                Ok(msg) => msg,
+                Err(try_rcv_err) => match try_rcv_err {
+                    crossbeam::channel::TryRecvError::Empty => continue,
+                    crossbeam::channel::TryRecvError::Disconnected => {
+                        return Err(BootstrapError::GeneralError(
+                            "listener recv channel disconnected unexpectedly".to_string(),
+                        )
+                        .into())
+                    }
                 },
+            };
 
-                // Whitelist cache timeout
-                _ = cache_interval.tick() => {
-                    (whitelist, blacklist) = reload_whitelist_blacklist(&self.bootstrap_config.bootstrap_whitelist_path, &self.bootstrap_config.bootstrap_blacklist_path)?;
-                }
+            // claim a slot in the max_bootstrap_sessions
+            let bootstrap_count_token = bootstrap_sessions_counter.clone();
+            let mut server = BootstrapServerBinder::new(
+                dplx,
+                self.keypair.clone(),
+                (&self.bootstrap_config).into(),
+            );
+
+            // check whether incoming peer IP is allowed.
+            // TODO: confirm error handled according to the previous `is_ip_allowed` fn
+            if let Err(error_msg) = self.allow_block_list.is_ip_allowed(&remote_addr) {
+                let _ = match self.runtime.block_on(server.send_error(error_msg.clone())) {
```

Comment:
is this the correct use of block_on?

Reply:
What do you means by correct ? You are waiting for the message to be sent. It's ok. I just find it strange to do all this check for allowance, cooldown in the main loop as we will take time to do it and take even more time to send the error and doing this time we are not doing anything.

If we follow something more like the architecture here : https://github.com/massalabs/massa/discussions/3149
The listener create threads for people that connect and it's on each of these thread that we verify the cooldown, allowance and send back the errors. So we don't block the main listening if a peer is not allowed take time to receive his error

Reply:
I'll spend some time now trying to move all that into the spawned thread.

Reply:
If we were to move everything into the handler thread, it would need mutable borrow of the server, which I really don't want to do at this stage as

a) I suspect that will cause a lot of issues on read/write contention
b) if we manage to put the error-sending off-thread (which I've got something almost done for that already), then the thread-spawning pre-amble should be acceptably quick.

At some stage, we will refine what gets moved into the thread, which might open the door to reducing the accept-to-spawn time

Reply:
I managed to move the error-sending stuff into a thread using the servers dedicated runtime.

Reply:
Ok but I think in the future it would be nice that the threads are spawn directly by the listener so that the main loop don't have to make all the verif things etc.. and we accept very fast. We can close this for now as it would be more appropriate to discuss in the thread PR.

Reply:
100% agree.

making the listener non-async is a bit non-trivial, as it ties in with the client code base and the mocks. 

I think making de-async task is coupled with the client. Will probably be much more efficient to do them together in another PR.

Reply:
Yes we can close this discussion